RCHAIN-3505: Error handling rholang

JosephDenman opened the pull request:
## Overview
A synchronous lock isn't used, but this is the ticket this approach grew from: <https://rchain.atlassian.net/browse/RCHAIN-3505?atlOrigin=eyJpIjoiZjE2YmNmNmU0NjY5NDllM2EzNWI1MzdmNjM0ZTBhNmYiLCJwIjoiaiJ9>

JosephDenman committed:
Simplify error reporting.

JosephDenman committed:
Rename `_error[F]` to `_bracket[F]`.

JosephDenman committed:
Introduce error reference alias and combinators.

JosephDenman committed:
Move `errors` into `error_handling` package.

JosephDenman committed:
Replace `ErrorLog` dependency with `_error` dependency.

JosephDenman committed:
Rename `readAndClearErrorVector` to `readAndClearErrorReference` and have it return `F[Option[Throwable]]`.

JosephDenman committed:
Use `error` to retrieve errors in `CostAccountingReducerTest`.

JosephDenman committed:
Use `_error[M]` to conditionally run processes and report errors in `Reduce`.

JosephDenman committed:
Make `parallelism` a local value.

JosephDenman committed:
Use `fixture` instead of `withTestSpace` in `ReduceSpec`.

JosephDenman committed:
Use `fixture` instead of `withRegistryAndTestSpace` in `RegistrySpec`.

JosephDenman committed:
Eliminate `withTestSpace` from `PersistentStoreTester`.

JosephDenman committed:
Have `fixture` take an optional error instead of `_error[Task]`.

JosephDenman commented:
bors try

JosephDenman committed:
Fix bug in `RhoSpec` contract.

JosephDenman commented:
bors try

JosephDenman committed:
Fix bug in integration tests.

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/error_handling/package.scala:
```diff
@@ -0,0 +1,25 @@
+      case Some(_) => ().pure[F]
+      case None =>
+        process.handleErrorWith { currentError =>
+          error.update {
```
this is easier to read and more compact: ```error.update(_.getOrElse(currentError))```

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/error_handling/package.scala:
```diff
@@ -0,0 +1,25 @@
+      process: F[Unit]
+  )(implicit error: _error[F]): F[Unit] =
+    error.get >>= {
+      case Some(_) => ().pure[F]
```
Instead of the match an if would be more readable:

```scala
error.get >>= (e =>
 if(e.isEmpty) ().pure[F]
else process.handleErrorWith {...} )
```

odeac commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/error_handling/package.scala:
```diff
@@ -0,0 +1,25 @@
+
+package object error_handling {
+
+  type _error[F[_]] = Ref[F, Option[Throwable]]
```
To me it feels that error is not a good name. An "error" should be a value, not a mutable storage.

I would go with _errorHandler instead of _error. Or maybe _errorStore

ArturGajowy replied:
The current name is fine to me. It won't matter though for reasons pointed out in my future comment.

ArturGajowy reviewed (changes requested):
Nice work! We need one more push (pun intended) or two before this is merged though.

1. The error ref and the `runAndReportErrors` method must be private to the DebriujinInterpreter, with the error created per call to `eval(Par)` and other entry points, and passed through the private methods of the reducer (implicit is fine for that). Otherwise it's just the global-state-ErrorLog in disguise, with less slots.
2. Reduction errors must be signalled by the Reducer via its effect type or its return type
3. We need to wait for the produce/consume-s in progress to finish before returning.
4. EvaluateResult should carry Option[Throwable] instead of a Vector

In effect:
 - the ```Replace `ErrorLog` dependency with `_error` dependency.``` will become ```Remove `ErrorLog```.
 - error_handling.ErrorHandling, and the error_handling packagbe object introduced in `Introduce error reference alias and combinators.` are going to disappear
 - the ```Move `errors` into `error_handling` package.``` will be redundant, and should be dropped

To make the above easier, it might make sense to:  
- Restrict the ChargingReducer interface to just `eval` and `inj`. `evalExpr` and `evalExprToPar` are only used in tests, and should be exposed to them via `private[rholang]`
- Restrict the Reduce interface to just `eval` and `inj`
- define the implementation of `inj` in the trait itself, or inline it whatsoever.

Make sure to remove the `ErrorLog` class, and any other code you're making unused.

The commits:
```
Have `fixture` take an optional error instead of `_error[Task]`. JosDenmark* 19.08.2019, 19:27
Eliminate `withTestSpace` from `PersistentStoreTester`. JosDenmark* 19.08.2019, 19:04
Use `fixture` instead of `withRegistryAndTestSpace` in `RegistrySpec`. JosDenmark* 19.08.2019, 19:03
Use `fixture` instead of `withTestSpace` in `ReduceSpec`. JosDenmark* 19.08.2019, 18:38
```
are a great improvement! They though should be squashed, as they edit the same places over and over again. Further, some the edited places still have a lot of duplicated structure, so it might make sense to push it a little more in there - and extract the whole repeated pattern at once instead of stripping it away bit by bit.  

~~```Make `parallelism` a local value.``` should be dropped, as there's no point in re-evaluating the `Reduce.parallelism` value over and over again. Feel free to make it private[rholang], or sth like that.~~

The `Fix bug in integration tests.` commit should be part of the commit that broke the tests.

ArturGajowy commented:
If possible, please prioritize the 'charge for system processes' work before coming back to this.

lukasz-golebiewski commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/error_handling/package.scala:
```diff
@@ -0,0 +1,25 @@
+import cats.MonadError
+import cats.effect.concurrent.Ref
+
+package object error_handling {
```
Why not introduce and nest the `handling` package under `error` and avoid the ugly underscore in the name?
I would even opt for getting rid of this package object completely and putting it's contents into the `error` package object

lukasz-golebiewski replied:
Just took a peek at the next commit. I find the name `error_handling.errors` confusing

JosephDenman replied:
It seems reasonable to keep the error classes and the error handler separate, no?

lukasz-golebiewski commented on rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala:
```diff
@@ -80,8 +81,8 @@ class ReduceSpec extends FlatSpec with Matchers with AppendedClues with Persiste
   "evalExpr" should "handle simple addition" in {
-    implicit val errorLog = new ErrorLog[Task]()
-    val result = withTestSpace(errorLog) {
+    implicit val error = ErrorHandling.emptyError[Task].runSyncUnsafe()
```
Adding more `runSyncUnsafe`'s decreases readability. Is it a lot of effort to get rid of them?

Does the `val error` need to be implicit if it's passed in explicitly on the line below?

JosephDenman replied:
No. It does not need to be implicit. But the next few commits eliminate `error` in the tests, so the correction is unnecessary.

lukasz-golebiewski replied:
What about the `runSyncUsafe`s?

JosephDenman replied:
I see what you mean. I'll remove them.

lukasz-golebiewski commented on rholang/src/test/scala/coop/rchain/rholang/interpreter/CostAccountingReducerTest.scala:
```diff
@@ -97,11 +97,12 @@ class CostAccountingReducerTest extends FlatSpec with Matchers with TripleEquals
   }
 
-  it should "stop interpreter threads as soon as deploy runs out of phlo" ignore {
+  it should "stop interpreter threads as soon as deploy runs out of phlo" in {
```
Yay!

lukasz-golebiewski commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:
```diff
@@ -184,8 +180,12 @@ class DebruijnInterpreter[M[_], F[_]](
       else if (terms.size > 256) rand.splitShort(id.toShort)
       else rand.splitByte(id.toByte)
 
+    // The number of processors available to the JVM may change over the lifetime of an application, so the number of
```
Good catch!

lukasz-golebiewski commented:
> Make `parallelism` a local value. should be dropped, as there's no point in re-evaluating the Reduce.parallelism value over and over again.

@ArturGajowy It turns out that the value of available processors for the JVM may change over time so Joe's change is justified

ArturGajowy commented:
Yup, I missed the comment :+1: That's the best "well, actually" I've seen in a while!

JosephDenman commented:
@ArturGajowy 

1. Basically, `error` must be in scope in `eval(par: Par)(...)` so that the error reference can be checked before the sub-processes are evaluated. This also prevents the duplication of `runAndReportErrors` in each of the evaluation methods. There are a few options:

- rewrite the `Reduce` interface so that `eval(par: Par)(...)` takes `error`.
- make a `create` method for `DebruijnInterpreter` that returns `F[DebruijnInterpreter[...]]`.
- keep the global dependency.

2. Fixed

3. This is enforced by the behavior of `parSequence_`. Previously, the evaluator used `raiseError` to cancel concurrent tasks - an inherently non-deterministic action. That is what was causing the produce/consume leakage before. We're no longer using `raiseError` to cancel concurrent tasks, so we sidestep those issues. This is easy to verify locally.

4.  Fixed
